File: Week_04/cpp/src/_79_WordSearch.cpp
#include <string>
#include <iostream>
#include <queue>

using namespace std;

class Dfs
{
public:
    vector<pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    // 时间 O(mn*3^L) 空间 O(min(L, mn)) L表示单词的长度
    bool exist(vector<vector<char>> &board, string word)
    {
        for (int i = 0; i < board.size(); i++)
        {
            for (int j = 0; j < board[0].size(); j++)
            {
                if (dfs(board, i, j, word, 0))
                    return true;
            }
        }
        return false;
    }
    bool dfs(vector<vector<char>> &board, int i, int j, string &path, int idx)
    {
        // if (idx == path.size()) return true;
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '#' || board[i][j] != path[idx++])
            return false;
        if (idx == path.size())
            return true;
        char c = board[i][j];
        board[i][j] = '#';
        bool ret = false;
        for (auto &[di, dj] : dirs)
        {
            if (dfs(board, i + di, j + dj, path, idx))
            {
                ret = true;
                break;
            }
        }
        board[i][j] = c;
        return ret;
    }
};

File: Week_04/kotlin/src/_79_WordSearch.kt
class _79_Dfs {
    // 时间 O(mn * 3^L) 空间 O(min(L, mn)) L表示word的长度
    fun exist(board: Array<CharArray>, word: String): Boolean {
        var idx = 0
        val n = word.length
        val dirs = arrayOf(intArrayOf(1, 0), intArrayOf(-1, 0), intArrayOf(0, 1), intArrayOf(0, -1))
        fun dfs(i: Int, j: Int): Boolean {
            if (idx == n) return true
            if (i !in board.indices || j !in board[0].indices || board[i][j] == '#' || board[i][j] != word[idx]) return false
            val c = board[i][j]
            board[i][j] = '#'
            idx++
            var ret = false
            for ((di, dj) in dirs) {
                if (dfs(i + di, j + dj)) {
                    ret = true
                    break
                }
            }
            idx--
            board[i][j] = c
            return ret
        }
        for (i in board.indices) {
            for (j in board[0].indices) {
                if (dfs(i, j)) return true
            }
        }
        return false
    }
}